### ISSUE TITLE
Linux Mock Server FFI Library Different to Windows

### ISSUE THREAD

**Issue by adamrodger** [pact-foundation/pact-reference#122]
I've downloaded the [latest version](https://github.com/pact-foundation/pact-reference/releases/tag/libpact_mock_server_ffi-v0.1.0) of the Mock Server FFI library for both Windows and Linux, and the Linux version is different. It doesn't export the `log_to_buffer` and other related logging commands that were added recently.

On Windows all of the functions are there:

![image](https://user-images.githubusercontent.com/508586/124976715-8a165d80-e027-11eb-9fc0-98f642444275.png)

But on Linux the newer log functions are missing:

![image](https://user-images.githubusercontent.com/508586/124976778-9b5f6a00-e027-11eb-953a-dd5dbfd159cf.png)

This is currently causing CI failures on PactNet when the Linux CI runs even though all of the tests pass locally on Windows.

Is it possible that the `pact-reference` CI is somehow uploading different versions of the libraries for the different OSs, or somehow building them differently so that the exported functions are different?

**Comment by adamrodger** [comment]
Just noticed they're not included in the [included C header file](https://github.com/pact-foundation/pact-reference/releases/download/libpact_mock_server_ffi-v0.1.0/pact_mock_server_ffi.h) either, which may have something to do with it. Perhaps re-exports don't work on some build types?

**Comment by uglyog** [comment]
They're built on different GH action agents, but the build process should be the same.

**Comment by uglyog** [comment]
I can confirm this happens locally, so this is an issue with the Rust build on Linux. Looks like functions can't be re-exported on Linux.

I've started the work to create a single FFI library, so that won't have this issue because there will be on re-exporting functions.

**Comment by adamrodger** [comment]
Are we missing a layer of (hopefully just basic) integration testing for
the FFI library which runs cross platform? It feels like the type of thing
we'd want to pick up early rather than post-release.

**Comment by uglyog** [comment]
There are tests against example projects (in C), but they don't test all exported functions and OSs, only run on Linux and using the mock server functions.

**Comment by TimothyJones** [comment]
> Are we missing a layer of (hopefully just basic) integration testing for the FFI library which runs cross platform?

This is a good point - we don't have a principled way of doing this at the moment. Pact-js has (kind of) integration tests, in that we run all the examples as step in the build. Doing this is actually really valuable and have picked up a few issues in the past, but they're not very "intentional" from an integration test coverage perspective.

Given that we're a contract testing framework, I'd like us to be more principled in the way we test the contract between the pact core and the wrappers, but I also don't see a straightforward way to do it. Do you have any thoughts?

**Comment by adamrodger** [comment]
> Do you have any thoughts?

If this is the reference implementation, then I think create a reference consumer of the reference FFI is the way forward, and then using a simple GitHub Actions automation to run that on all platforms. It doesn't really matter which language that's written in really as long as we test the FFI on all target platforms.

I definitely don't think we should be in a situation where we find the a certain OS/feature doesn't work only once we've published a version of the FFI and downstream integrators have started using it.

Anyway, this issue itself can now be closed because the original issue no longer matters with the new combined FFI, but I'll open a new issue to make sure we're running CI across all platforms. PactNet runs on Windows, Linux and Mac at present using GitHub Actions, so the FFI builds should be able to do the same.
